kzg_prover/bin/gen_verifiers.rs

```diff
@@ -0,0 +1,188 @@
+        .unwrap();
+
+    let univariate_grand_sum_circuit =
+        UnivariateGrandSum::<N_USERS, N_CURRENCIES>::init(entries.to_vec());
+
+    // 1. Generate Snark Verifier Contract and Verification Key
+    //
+    // the instance value is not used in proving, but it is necessary to SolidityGenerator and it should at least 1.
+    let num_instance = 1_usize;
+    let generator: SolidityGenerator<'_> =
+        SolidityGenerator::new(&params, pk.get_vk(), Bdfg21, num_instance);
+    let (verifier_solidity, vk_verifier) = generator.render_separately().unwrap();
+
+    let verifier_solidity_fixed = verifier_solidity
+        .replace("Halo2Verifier", "Verifier")
+        .replace(") public returns (bool)", ") public view returns (bool)");
+
+    let verifier_code = compile_solidity(&verifier_solidity_fixed);
+    let vk_code = compile_solidity(&vk_verifier);
+
+    // 2. Generate Snark Proof for range check
+    //
+    // the instance values has to be at least more than one due to verifier contract that generated from SolidityGenerator.
+    let instances: Vec<Fr> = vec![Fp::one(); 1];
+    let mut transcript = Keccak256Transcript::new(Vec::new());
+
+    let result = create_proof::<_, ProverSHPLONK<_>, _, _, _, _>(
+        &params,
+        &pk,
+        &[univariate_grand_sum_circuit],
+        &[&[&instances]],
+        &mut OsRng,
+        &mut transcript,
+    );
+    assert!(result.is_ok());
+
+    let result_unwrapped = result.unwrap();
+    result_unwrapped.0.expect("prover should not fail");
+    let _advice_polys = result_unwrapped.1.clone();
+    let advice_polys = _advice_polys[0].clone();
+
+    let omega = pk.get_vk().get_domain().get_omega();
+    let zk_snark_proof = transcript.finalize();
+
+    // Check verification on verifier function
+    let verified = {
+        let mut transcript = Keccak256Transcript::new(zk_snark_proof.as_slice());
+        verify_proof::<_, VerifierSHPLONK<_>, _, _, _>(
+            &params,
+            pk.get_vk(),
+            SingleStrategy::new(&params),
+            &[&[&instances]],
+            &mut transcript,
+        )
+    };
+    assert!(verified.is_ok());
+
+    // 3. Deploy Snark Verifier Contract and verify snark proof
+    let mut evm = Evm::default();
+
+    // Calldata for verifying proof on evm
+    let vk_address = evm.create(vk_code);
+    let proof_calldata = encode_calldata(Some(vk_address.into()), &zk_snark_proof, &instances);
+
+    // Initiate verifier contract
+    let verifier_address = evm.create(verifier_code);
+    let (_, output) = evm.call(verifier_address, proof_calldata);
+
+    // If successfuly verified, the verifier contract will return 1.
+    assert_eq!(output, [vec![0; 31], vec![1]].concat());
+    save_solidity("snark_verifier.sol", &verifier_solidity_fixed);
+    save_solidity("verifying_key.sol", &vk_verifier);
+
+    // 4. Generate Opening Proof for Grand Sums
```

Comment:
I think it's ok to remove the code. related to the grand sum opening until we have an actual verifier for that, otherwise this code acts like a duplicated unit test for the grand sum opening because the result is not used anywhere.

Reply:
Okay, I will remove that now.